src/core/or/circuitpadding.h

```diff
@@ -0,0 +1,539 @@
+  uint8_t chosen_bin;
+
+  /** Stop padding/transition if this many cells sent */
+  uint64_t state_length;
+#define CIRCPAD_STATE_LENGTH_INFINITE UINT64_MAX
+
+  /** A scaled count of padding packets sent, used to limit padding overhead.
+   * When this reaches UINT16_MAX, we cut it and nonpadding_sent in half. */
+  uint16_t padding_sent;
+  /** A scaled count of non-padding packets sent, used to limit padding
+   *  overhead. When this reaches UINT16_MAX, we cut it and padding_sent in
+   *  half. */
+  uint16_t nonpadding_sent;
+
+  /**
+   * EWMA estimate of the RTT of the circuit from this hop
+   * to the exit end, in microseconds. */
+  circpad_delay_t rtt_estimate_us;
+
+  /**
+   * The last time we got an event relevant to estimating
+   * the RTT. Monotonic time in microseconds since system
+   * start.
+   */
+  circpad_time_t last_received_time_us;
+
+  /**
+   * The time at which we scheduled a non-padding packet,
+   * or selected an infinite delay.
+   *
+   * Monotonic time in microseconds since system start.
+   * This is 0 if we haven't chosen a padding delay.
+   */
+  circpad_time_t padding_scheduled_at_us;
+
+  /** What state is this machine in? */
+  ENUM_BF(circpad_statenum_t) current_state : 2;
+
+  /**
+   * True if we have scheduled a timer for padding.
+   *
+   * This is 1 if a timer is pending. It is 0 if
+   * no timer is scheduled. (It can be 0 even when
+   * padding_was_scheduled_at_us is non-zero).
+   */
+  uint8_t padding_timer_scheduled : 1;
+
+  /**
+   * If this is true, we have seen full duplex behavior.
+   * Stop updating the RTT.
+   */
+  uint8_t stop_rtt_update : 1;
+
+/** Max number of padding machines on each circuit. If changed,
+ * also ensure the machine_index bitwith supports the new size. */
+#define CIRCPAD_MAX_MACHINES    (2)
+  /** Which padding machine index was this for.
+   * (make sure changes to the bitwidth can support the
+   * CIRCPAD_MAX_MACHINES define). */
+  uint8_t machine_index : 1;
+
+} circpad_machineinfo_t;
+
+/** Helper macro to get an actual state machine from a machineinfo */
+#define CIRCPAD_GET_MACHINE(machineinfo) \
+    ((machineinfo)->on_circ->padding_machine[(machineinfo)->machine_index])
+
+/**
+ * This specifies a particular padding machine to use after negotiation.
+ *
+ * The constants for machine_num_t are in trunnel.
+ * We want to be able to define extra numbers in the consensus/torrc, though.
+ */
+typedef uint8_t circpad_machine_num_t;
+
+/** Global state machine structure from the consensus */
+typedef struct circpad_machine_t {
+  /** Global machine number */
+  circpad_machine_num_t machine_num;
+
+  /** Which machine index slot should this machine go into in
+   *  the array on the circuit_t */
+  uint8_t machine_index : 1;
+
+  /** Send a padding negotiate to shut down machine at end state? */
+  uint8_t negotiate_end : 1;
+
+  // These next three fields are origin machine-only...
+  /** Origin side or relay side */
+  uint8_t origin_side : 1;
```

Comment:
> Also, we can turn this into a `bool` instead of a `uint8_t :1`. Same for all the booleans above. I think it might be clearer.

Careful: 3 bools take up 3 bytes, but 3 bitfields take up 3 bits, and those 3 bits will will be packed into the same byte as the 3-bit target_hopnum field.

Reply:
I don't think there are enough `circpad_machine_t` structs alive to worry about sizes here, but I agree that this might be suboptimal for not much gain. Let's forget turning them to bools.

However, changing the variable names to indicate they are booleans is still worth it IMO.

Reply:
All the variable changing you guys asked for in this and other comments is done across these three commits (so as to squash sanely): d8c4eff8c5d62d172eac0d5096f32f0be99ce6cf 979cc31a3e3e97619760c7aa4f94c8f767cdd29a e18d5a3111375b995a57e809f080011d6ebe5ad5